### PR TITLE
only forward inputs with one use to avoid introducing additional persistent buffer

### PR DESCRIPTION
**Pull request by liqiangxl**
See https://github.com/NVIDIA/Fuser/issues/1607

### CODE REVIEW

**Comment by liqiangxl** [comment]
!build --diff --diff-bench

**Comment by liqiangxl** [comment]
!build --diff --diff-bench

**Comment by naoyam** [comment]
> Just a comment. Looking at [this diff](http://nv/e5M/nvfuser_github_ci/codegen_diff_p12046756_j79498625_1705030929968697353_codediff_7c8b098_931601a_custom_command_20240112_032916.html) it looks like the test for #1301 is reduced from three segments to two with this PR (test is misnamed, oops). The other diffs are just reordering. That test contains three reductions and on main it is producing three separate reduction kernels, two grid reductions and a block reduce, while in this PR both the grid reductions are put into a single reduction kernel which is probably good.

I think that's because stripping out multi-use inputs would make a DAG disjoint, so we would end up having separate segments. 

Another side effect of excluding multi-use inputs is what happened with the original bug, i.e., missing to detect a normalization pattern. 

@liqiangxl This change makes sense to me. Please add more comments to the code why we shouldn't exclude multi-use inputs.

File: csrc/fusion_segmenter.cpp [hunk 1]
```diff
@@ -3779,14 +3779,17 @@ void SegmentCandidateFinder::forwardInputs() {
...
+      // Stripping out multi-used inputs can lead to:
+      // (1) Fragmentation of the DAG, increased segments, see test in #1301.
+      // (2) Miss detection of persistent buffers, see issue #1607.
+      const auto& input_uses = inp->uses();
+      if (input_uses.size() != 1) {
+        continue;
+      }
+      // Add the use of input if it is a UnaryOp
+      if (input_uses.at(0)->isA<UnaryOp>()) {
+        to_visit.push_back(input_uses.at(0)->as<UnaryOp>());
```
Comment (jacobhinkle) [inline csrc/fusion_segmenter.cpp:3792]:
```suggestion
      // Add single-use input if it is a UnaryOp
      if (input_uses.size() == 1 && input_uses.at(0)->isA<UnaryOp>()) {
        to_visit.push_back(input_uses.at(0)->as<UnaryOp>());
```

**Review by jacobhinkle — APPROVED** [review]
LGTM

**Pull request merged**